Support alternate Drupal distributions

grayside opened the pull request:
The changes below are a fairly significant refactoring of the code to allow individual distributions to take ownership of relevant options presented to the user and the specific operations used to assemble the distribution.

As a case, we have "vanilla" Drupal and Open Atrium in this change. As far as I can tell, both are correctly setting up the codebase and the main Grunt process works fine.
- Drupal 8 is broken by #10 
- Drupal 7 successfully ran with `grunt drush:liteinstall && grunt serve`
- Atrium through an undefined index warning on the "quickstart" variable, and then hit a fatal SQL exception:

```
exception 'PDOException' with message 'SQLSTATE[HY000]: General error: 1 table "defaultconfig_optionals" has more than one primary key' in                              [error]
[...]/tools/hold/build/html/includes/database/sqlite/database.inc:238
Stack trace:
```

This fixes #2

@arithmetric 
@mike-potter

Commit:
Add support for multiple distributions via plugin mechanism.

Commit:
Correct Atrium makefile assembly.

Commit:
Specify Drupal profile for grunt install and serve tasks.

arithmetric commented on app/distros/index.js:
```diff
@@ -0,0 +1,2 @@
+var requireDirectory = require('require-directory');
```
This module needs to be added to the package.json.

arithmetric commented:
When I run this and choose any version of Drupal or Atrium, I get the following error:

```
? Which Drupal distribution would you like to use? Drupal
? Which version of Drupal would you like to use? Drupal 7

Ok, I'm going to start assembling this project...

events.js:72
        throw er; // Unhandled 'error' event
              ^
TypeError: Cannot read property 'dependencies' of undefined
    at module.exports.yeoman.generators.Base.extend.writing.packageJson (/Users/Turgeon/work/automation/grush/generator-gadget/app/index.js:120:12)
    at /Users/Turgeon/work/automation/grush/generator-gadget/node_modules/yeoman-generator/lib/base.js:409:16
    at processImmediate [as _immediateCallback] (timers.js:363:15)
```

arithmetric commented on app/index.js:
```diff
@@ -42,69 +59,31 @@ module.exports = yeoman.generators.Base.extend({
-      });
-    });
+  installGDT: function() {
+    //require('./gdt')(this).install();
```
Maybe this is why there is no package.json?

arithmetric commented:
Overall, this looks like a good direction. :+1:

I've noted a few issues when testing it.

Commit:
Added missing pieces that were breaking the generator.

Commit:
Revised default makefiles and set up a generic makefile for Drupal.org hosted distros to share.

Commit:
Add test coverage to distro plugins and adjust existing test to match new parameters.

grayside commented:
1. Added require-directory to package.json
2. Uncommented gdt installation
3. Standardized minor formatting and default details of makefiles.
4. Move openatrium project.make template to templates/ for reuse across D.O distros.
5. Added test coverage to verify compliance of distro plugins.
6. Adjusted existing test to match new option parameter names (could not get this test to run as it seems to have some specific environment requirements I'm not aware of)

This change does not fix the openatrium DB error noted in the description above, but I think that should be addressed as a follow-up.
